File: studio/repositories/versions_repo.py
"""Version snapshot persistence."""

import hashlib
import time
import uuid

from studio import db
from studio.models import Version, to_doc


def _profile_hash(profile):
    raw = str(sorted((profile or {}).items()))
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:16]


def create(project_id, *, label, trigger, html, profile=None):
    version_id = uuid.uuid4().hex[:12]
    now = time.time()
    row = Version(
        versionId=version_id,
        projectId=project_id,
        label=label,
        trigger=trigger,
        createdAt=now,
        snapshotHtml=html,
        snapshotProfileHash=_profile_hash(profile),
        s3HtmlKey=f"projects/{project_id}/versions/{version_id}.html",
    )
    with db.session() as s:
        s.add(row)
        s.commit()
        return to_doc(row)


def list_for_project(project_id):
    with db.session() as s:
        rows = (
            s.query(Version)
            .filter_by(projectId=project_id)
            .order_by(Version.createdAt.desc())
            .all()
        )
        return [to_doc(r) for r in rows]


def get(project_id, version_id):
    with db.session() as s:
        row = (
            s.query(Version)
            .filter_by(projectId=project_id, versionId=version_id)
            .first()
        )
        return to_doc(row)


def head(project_id):
    versions = list_for_project(project_id)
    return versions[0] if versions else None

File: src/server/repos/versionsRepo.ts
import { createHash } from "node:crypto";
import { prisma } from "../db";
import { toDoc } from "./doc";

function profileHash(profile: Record<string, unknown> | null | undefined): string {
  const raw = JSON.stringify(Object.entries(profile ?? {}).sort());
  return createHash("sha256").update(raw).digest("hex").slice(0, 16);
}

export async function create(
  projectId: string,
  data: {
    label: string;
    trigger: string;
    html: string;
    profile?: Record<string, unknown> | null;
  },
): Promise<Record<string, unknown>> {
  const versionId = randomId();
  const now = Date.now() / 1000;
  const row = await prisma().builderVersion.create({
    data: {
      versionId,
      projectId,
      label: data.label,
      trigger: data.trigger,
      createdAt: now,
      snapshotHtml: data.html,
      snapshotProfileHash: profileHash(data.profile),
      s3HtmlKey: `projects/${projectId}/versions/${versionId}.html`,
    },
  });
  return toDoc(row) as Record<string, unknown>;
}

export async function listForProject(projectId: string): Promise<Record<string, unknown>[]> {
  const rows = await prisma().builderVersion.findMany({
    where: { projectId },
    orderBy: { createdAt: "desc" },
  });
  return rows.map((row) => toDoc(row) as Record<string, unknown>);
}

export async function get(
  projectId: string,
  versionId: string,
): Promise<Record<string, unknown> | null> {
  const row = await prisma().builderVersion.findFirst({ where: { projectId, versionId } });
  return toDoc(row) as Record<string, unknown> | null;
}

export async function head(projectId: string): Promise<Record<string, unknown> | null> {
  const versions = await listForProject(projectId);
  return versions[0] ?? null;
}

function randomId(): string {
  const buf = new Uint8Array(6);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}
